packages/core/src/hydration/event_replay.ts

```diff
@@ -0,0 +1,173 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Dispatcher, EventContract, EventInfoWrapper, registerDispatcher} from '@angular/core/primitives/event-dispatch';
+
+import {APP_BOOTSTRAP_LISTENER, ApplicationRef, whenStable} from '../application/application_ref';
+import {APP_ID} from '../application/application_tokens';
+import {Injector} from '../di';
+import {inject} from '../di/injector_compatibility';
+import {Provider} from '../di/interface/provider';
+import {getLContext} from '../render3/context_discovery';
+import {TNode, TNodeType} from '../render3/interfaces/node';
+import {RNode} from '../render3/interfaces/renderer_dom';
+import {CLEANUP, LView, TVIEW, TView} from '../render3/interfaces/view';
+import {isPlatformBrowser} from '../render3/util/misc_utils';
+import {unwrapRNode} from '../render3/util/view_utils';
+
+import {IS_EVENT_REPLAY_ENABLED} from './tokens';
+
+export const EVENT_REPLAY_ENABLED_DEFAULT = false;
+export const CONTRACT_PROPERTY = 'ngContracts';
+
+declare global {
+  var ngContracts: {[key: string]: EventContract};
+}
+
+/**
+ * Returns a set of providers required to setup support for event replay.
+ * Requires hydration to be enabled separately.
+ */
+export function withEventReplay(): Provider[] {
+  return [
+    {
+      provide: IS_EVENT_REPLAY_ENABLED,
+      useValue: true,
+    },
+    {
+      provide: APP_BOOTSTRAP_LISTENER,
+      useFactory: () => {
+        if (isPlatformBrowser()) {
+          const injector = inject(Injector);
+          const appRef = inject(ApplicationRef);
+          return () => {
+            whenStable(appRef).then(() => {
+              const appId = injector.get(APP_ID);
+              // This is set in packages/platform-server/src/utils.ts
+              const eventContract = globalThis[CONTRACT_PROPERTY][appId] as EventContract;
+              if (eventContract) {
+                const dispatcher: Dispatcher = new Dispatcher();
+                setEventReplayer(appRef, new Set([appId]), dispatcher);
+                registerDispatcher(eventContract, dispatcher);
+              }
+            });
+          };
+        }
+        return () => {};  // noop
```

Comment:
nit: we can expand the comment to mention that this is a noop for the server code:
```suggestion
        return () => {};  // noop for the server code
```

Reply:
done